receiver/prometheusreceiver/internal/staleness_store.go

```diff
@@ -0,0 +1,78 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package internal
+
+import (
+	"github.com/prometheus/prometheus/pkg/labels"
+)
+
+type stalenessStore struct {
+	currentHashes  map[uint64]bool
+	previousHashes map[uint64]bool
+	previous       []labels.Labels
+	current        []labels.Labels
+}
+
+func newStalenessStore() *stalenessStore {
+	return &stalenessStore{
+		previousHashes: make(map[uint64]bool),
+		currentHashes:  make(map[uint64]bool),
+	}
+}
+
+// refresh copies over all the current values to previous, and prepares.
+func (ss *stalenessStore) refresh() {
+	// 1. Clear ss.previousHashes firstly. Please don't edit
+	// this map clearing idiom as it ensures speed.
```

Comment:
If you want to ensure performance doesn't regress, it might be useful to add a benchmark.

Reply:
It's more of Go compiler magic to let future code maintainers to know the importance of it as few people know about this. Adding benchmarks won't add much value honestly, the comment does suffice IMHO.

Reply:
Is there a reference you can give to explain the optimization? Not every maintainer is knowledgeable about this and would be useful to have a reference.

Reply:
Done, please take a look.